File: agent/commands-generator/src/main/java/org/matilda/commands/python/writer/PythonCodeBlock.java
package org.matilda.commands.python.writer;

public class PythonCodeBlock {
    private enum ElementKind {
        NONE,
        STATEMENT,
        BLOCK,
    }
    private static final int NORMAL_EMPTY_LINES = 2;
    private static final int CLASS_EMPTY_LINES = 1;

    private final CodeBlock mCodeBlock;
    private final int mEmptyLinesBetweenStuff;
    private ElementKind mLastElementKind;

    public PythonCodeBlock(CodeBlock codeBlock) {
        this(codeBlock, NORMAL_EMPTY_LINES);
    }
    private PythonCodeBlock(CodeBlock codeBlock, int emptyLines) {
        mCodeBlock = codeBlock;
        mEmptyLinesBetweenStuff = emptyLines;
        mLastElementKind = ElementKind.NONE;
    }

    public PythonCodeBlock newFunction(PythonFunctionSpec function) {
        addEmptyLinesBeforeBlock();
        mCodeBlock.addStatement("%s:", function.getDeclaration());
        return createNewCodeBlockWithIndentation();
    }

    private PythonCodeBlock createNewCodeBlockWithIndentation() {
        mLastElementKind = ElementKind.BLOCK;
        return new PythonCodeBlock(mCodeBlock.newCodeBlockWithIndentation());
    }

    public PythonCodeBlock addStatement(String statement) {
        addEmptyLinesBeforeStatement();
        mLastElementKind = ElementKind.STATEMENT;
        mCodeBlock.addStatement(statement);
        return this;
    }

    public PythonCodeBlock addStatement(String format, Object... args) {
        mCodeBlock.addStatement(format, args);
        return this;
    }

    private void addEmptyLinesBeforeStatement() {
        if (mLastElementKind == ElementKind.BLOCK) {
            addEmptyLines();
        }
    }

    private void addEmptyLinesBeforeBlock() {
        if (mLastElementKind != ElementKind.NONE) {
            addEmptyLines();
        }
    }

    private void addEmptyLines() {
        for (int i = 0; i < mEmptyLinesBetweenStuff; i++) {
            mCodeBlock.addEmptyLine();
        }
    }
}

File: agent/commands-generator/src/main/java/org/matilda/commands/python/writer/PythonCodeBlock.kt
package org.matilda.commands.python.writer

import java.util.stream.Stream

open class PythonCodeBlock protected constructor(
    private val mCodeBlock: CodeBlock,
    private val mEmptyLinesBetweenStuff: Int
) {
    private enum class ElementKind {
        NONE,
        STATEMENT,
        BLOCK
    }

    private var mLastElementKind = ElementKind.NONE

    constructor(codeBlock: CodeBlock = CodeBlock()) : this(codeBlock, NORMAL_EMPTY_LINES)

    fun newFunction(function: PythonFunctionSpec): PythonCodeBlock {
        addSeparatorBeforeBlock()
        addAnnotations(function.annotations)
        mCodeBlock.addStatement("%s:", function.declaration)
        return PythonCodeBlock(createNewCodeBlockWithIndentation())
    }

    private fun addAnnotations(annotations: List<String>) {
        annotations.forEach { annotation -> addStatement("@%s", annotation) }
    }

    fun newClass(classSpec: PythonClassSpec): PythonClass {
        addSeparatorBeforeBlock()
        mCodeBlock.addStatement("%s:", classSpec.declaration)
        return PythonClass(createNewCodeBlockWithIndentation())
    }

    private fun createNewCodeBlockWithIndentation(): CodeBlock {
        mLastElementKind = ElementKind.BLOCK
        return mCodeBlock.newCodeBlockWithIndentation()
    }

    fun addStatement(statement: String): PythonCodeBlock {
        addSeparatorBeforeStatement()
        mLastElementKind = ElementKind.STATEMENT
        mCodeBlock.addStatement(statement)
        return this
    }

    fun addStatement(format: String, vararg args: Any): PythonCodeBlock {
        mCodeBlock.addStatement(format, *args)
        return this
    }

    private fun addSeparatorBeforeStatement() {
        if (mLastElementKind == ElementKind.BLOCK) {
            addSeparator()
        }
    }

    private fun addSeparatorBeforeBlock() {
        if (mLastElementKind != ElementKind.NONE) {
            addSeparator()
        }
    }

    private fun addSeparator() {
        repeat (mEmptyLinesBetweenStuff) {
            mCodeBlock.addEmptyLine()
        }
    }

    open val lines: Stream<String>
        get() = mCodeBlock.lines

    companion object {
        private const val NORMAL_EMPTY_LINES = 2
    }
}
